libraries/entities-renderer/src/RenderableModelEntityItem.cpp

```diff
@@ -1460,6 +1460,31 @@ void ModelEntityRenderer::doRenderUpdateSynchronousTyped(const ScenePointer& sce
         emit requestRenderUpdate();
     }
 
+    if (!_allProceduralMaterialsLoaded) {
+        std::lock_guard<std::mutex> lock(_materialsLock);
+        bool allProceduralMaterialsLoaded = true;
+        for (auto& shapeMaterialPair : _materials) {
+            auto material = shapeMaterialPair.second;
+            while (!material.empty()) {
+                auto mat = material.top();
+                if (mat.material && mat.material->isProcedural() && !mat.material->isReady()) {
+                    allProceduralMaterialsLoaded = false;
+                    break;
+                }
+                material.pop();
+            }
+            if (!allProceduralMaterialsLoaded) {
+                break;
+            }
+        }
+        if (!allProceduralMaterialsLoaded) {
+            emit requestRenderUpdate();
```

Comment:
I think this signal is called with the lock held, is that intentional?

Reply:
yep, that’s fine!

Reply:
Mm, ok then. My understanding is that Qt invokes a signal immediately if the receiver is in the same thread, so the lock would be held however long the handler takes, which seems undesirable.

However upon checking I suppose in this particular case it doesn't matter much, since the one signal handler there is uses a queued connection.